core/shared/src/main/scala/zio/ZIO.scala

```diff
@@ -1763,12 +1763,17 @@ private[zio] trait ZIOFunctions extends Serializable {
   final def foreachParN[R, E, A, B](
     n: Long
   )(as: Iterable[A])(fn: A => ZIO[R, E, B]): ZIO[R, E, List[B]] =
-    (for {
-      semaphore <- Semaphore.make(n)
-      bs <- foreachPar[R, E, A, B](as) { a =>
-             semaphore.withPermit(fn(a))
-           }
-    } yield bs).refailWithTrace
+    for {
+      q     <- Queue.bounded[(Promise[E, B], A)](n.toInt)
+      pairs <- ZIO.foreach(as)(a => Promise.make[E, B].map(p => (p, a)))
+      _     <- ZIO.foreach(pairs)(pair => q.offer(pair)).fork
+      _ <- ZIO.collectAll(
+            List.fill(n.toInt)(
+              q.take.flatMap { case (p, a) => fn(a).foldCauseM(e => p.halt(e) *> q.shutdown, b => p.succeed(b)) }.forever.fork
```

Comment:
I wonder if the correct thing to do here is to halt all the promises, rather than just the current one. `q.shutdown` doesn't seem to reliably interrupt the other fibers the way I thought it would. The following seems to work a lot better

`fn(a).foldCauseM(c => ZIO.foreach(promises)(_._2.halt(c)), p.succeed)`

This seems to handle interruption better.

Reply:
Is there an example you're looking at where you're seeing this?

Reply:
If you have a list like

List(sleep10sAndThenFail, succeed, failDirectly)

I would expect that to fail as soon as it's evaluated by foreachParN (given n > 2). In my testing it didn't actually fail directly, it needed to wait for 10s before it failed. It's possible I messed up my test (away from my computer at the moment).

Reply:
Fixed now. I think the issue was that previous approach was imposing a sequencing on the failures. So if we had `List(ZIO.fail("A"), ZIO.fail("B"))` we were guaranteed to get `ZIO.fail("A")`, which is great for determinism but means we have to wait for the first failure in the list if it takes longer than the other failure. In your example, if you switched the ordering of sleep10sAndThenFail and failDirectly it returns immediately.